plugins/inlineStyles.js

```diff
@@ -0,0 +1,181 @@
+  var styleEls  = document.querySelectorAll('style');
+
+  //no <styles/>s, nothing to do
+  if(styleEls === null) {
+    return document;
+  }
+
+  var styles    = [],
+      selectors = [];
+
+  for(var styleEl of styleEls) {
+    if(styleEl.isEmpty()) {
+      // skip empty <style/>s
+      continue;
+    }
+    var cssStr = styleEl.content[0].text || styleEl.content[0].cdata || [];
+
+    // collect <style/>s and their css ast
+    var cssAst = csstree.parse(cssStr, {context: 'stylesheet'});
+    styles.push({
+      styleEl: styleEl,
+      cssAst:  cssAst
+    });
+
+    selectors = selectors.concat(cssTools.flattenToSelectors(cssAst));
+  }
+
+
+  // filter for mediaqueries to be used or without any mediaquery
+  var selectorsMq = cssTools.filterByMqs(selectors, opts.useMqs);
+
+
+  // filter for pseudo elements to be used
+  var selectorsPseudo = cssTools.filterByPseudos(selectorsMq, opts.usePseudos);
+
+  // remove PseudoClass from its SimpleSelector for proper matching
+  cssTools.cleanPseudos(selectorsPseudo);
+
+
+  // stable sort selectors
+  var sortedSelectors = cssTools.sortSelectors(selectorsPseudo).reverse();
+
+
+  // apply <style/> styles to matched elements
+  for(var selector of sortedSelectors) {
+    var selectorStr = csstree.translate(selector.item.data),
+        selectedEls = null;
+
+    try {
+        selectedEls = document.querySelectorAll(selectorStr);
+    } catch(e) {
+        if(e.constructor == SyntaxError) {
+            console.warn('Syntax error when trying to select \n\n' + selectorStr + '\n\n, skipped.');
+            continue;
+        }
+        throw e;
+    }
+
+    if(selectedEls === null) {
+      // nothing selected
+      continue;
+    }
+
+    if(opts.onlyMatchedOnce && selectedEls !== null && selectedEls.length > 1) {
+      // skip selectors that match more than once if option onlyMatchedOnce is enabled
+      continue;
+    }
+
+    // apply <style/> to matched elements
+    for(var selectedEl of selectedEls) {
+
+      if(selector.rule === null) {
+        continue;
+      }
+
+      // merge declarations
+      csstree.walkDeclarations(selector.rule, function(styleCssoDeclaration) {
+        var styleDeclaration = cssTools.cssoToStyleDeclaration(styleCssoDeclaration);
+        if(selectedEl.style.getPropertyValue(styleDeclaration.name)    !== null &&
+           selectedEl.style.getPropertyPriority(styleDeclaration.name) >=  styleDeclaration.property.priority) {
+          return;
+        }
+        selectedEl.style.setProperty(styleDeclaration.name, styleDeclaration.property.value, styleDeclaration.property.priority);
+      });
+
+    }
+
+    if(opts.removeMatchedSelectors && selectedEls !== null && selectedEls.length > 0) {
+      // clean up matching simple selectors if option removeMatchedSelectors is enabled
+      selector.rule.selector.children.remove(selector.item);
+    }
+  }
+
+
+  for(var style of styles) {
+    csstree.walkRules(style.cssAst, function(node, item, list) {
```

Comment:
This was considered in https://github.com/svg/svgo/pull/592#issuecomment-244979133.
For cleaning up empty <style/>s, the empty Atrules have to be cleaned up first, too.